common-libs/common/src/Affinity.ts

```diff
@@ -600,7 +601,7 @@ export class Affinity {
 
   async deriveSegmentProof<TKeys extends string, TData extends SimpleThing & Record<TKeys, unknown>>(
     credential: VCV1<VCV1Subject<TData>>,
-    fields: TKeys[],
+    fields: string[],
```

Comment:
Could you please explain this change?
And `TKeys` generic parameter was only needed to put constraints on `credential` and `fields` relationship (`fields` should only contain a subset of non-simple-thing keys in `TData`). If `fields` can contain any strings now, we no longer need `TKeys` parameter at all, we don't even need `TData` probably and can simply write `credential: VCV1<VCV1Subject<SimpleThing>>`. If that's really what we want to do.

Reply:
If we want fields to contain complex paths, we could do it in a type-safe fashion like that (just to give an idea):
```
type InnerPaths<TObject extends Record<string, unknown>> = TObject extends Record<string, unknown> ? {
  // @ts-ignore
  [key in keyof TObject]: key extends string ? `${key}.${Paths<TObject[key]>}` : never
} : never

type Paths<TObject> = TObject extends Record<string, unknown>
    ? ((keyof TObject) | InnerPaths<TObject>[keyof InnerPaths<TObject>])
    : never

async deriveSegmentProof<TData extends SimpleThing>(credential: VCV1<VCV1Subject<TData>>, fields: Paths<TData>) { ... }
```
![image](https://user-images.githubusercontent.com/82827564/145414792-0ffad678-80fb-4b96-bc30-773a4b65ff48.png)

Or just get rid of generic parameters altogether.

Reply:
Template literals, ok, let's try them. Don't see any objections

Reply:
It leads to fails during [dependent lib compilation](https://github.com/affinityproject/affinidi-core-sdk/runs/4471050943?check_suite_focus=true)
Also, I can't manage find a way how to cover VCs with nested arrays, e.g.
<details>
  <summary>Example of VC with Arrays (probably HAT)</summary>

```JSON
{
    "type": "HealthPassportBundleCredentialV1",
    "data": {
        "@type": "BundleContainer",
        "fhirVersion": "4.0.1",
        "fhirBundle": {
            "resourceType": "Bundle",
            "@type": "Bundle",
            "entry": [
                {
                    "resource": {
                        "resourceType": "Immunization",
                        "patient": {
                            "reference": "Patient/0"
                        },
                        "vaccineCode": "COVID-19 vaccine, inactivated virus",
                        "manufacturer": {
                            "reference": "Organization/manufacturer-0"
                        },
                        "lotNumber": "SINO-100",
                        "occurrenceDateTime": "2021-06-17T11:30:00+08:00",
                        "doseQuantity": {
                            "value": 1
                        },
                        "performer": {
                            "actor": {
                                "reference": "Organization/administeringCentre-0"
                            }
                        },
                        "protocolApplied": {
                            "seriesDosesPositiveInt": 2
                        }
                    }
                },
                {
                    "resource": {
                        "resourceType": "Immunization",
                        "patient": {
                            "reference": "Patient/0"
                        },
                        "vaccineCode": "COVID-19 vaccine, RNA based",
                        "manufacturer": {
                            "reference": "Organization/manufacturer-1"
                        },
                        "lotNumber": "PB-100",
                        "occurrenceDateTime": "2021-06-17T11:30:00+08:00",
                        "doseQuantity": {
                            "value": 1
                        },
                        "performer": {
                            "actor": {
                                "reference": "Organization/administeringCentre-1"
                            }
                        },
                        "protocolApplied": {
                            "seriesDosesPositiveInt": 2
                        }
                    }
                },
                {
                    "resource": {
                        "resourceType": "Patient",
                        "id": "0",
                        "identifier": [
                            {
                                "value": "S1234567A",
                                "type": {
                                    "text": "National ID"
                                }
                            },
                            {
                                "value": "K1234567Z",
                                "type": {
                                    "text": "Passport ID"
                                }
                            }
                        ],
                        "name": [
                            {
                                "text": "Alice"
                            }
                        ],
                        "gender": "Female",
                        "birthDate": "1990-12-31"
                    }
                },
                {
                    "resource": {
                        "resourceType": "Organization",
                        "id": "manufacturer-0",
                        "name": "Sinovac"
                    }
                },
                {
                    "resource": {
                        "resourceType": "Organization",
                        "id": "manufacturer-1",
                        "name": "Pfizer-BioNTech"
                    }
                },
                {
                    "resource": {
                        "resourceType": "Organization",
                        "id": "administeringCentre-0",
                        "name": "Outram Polyclinic",
                        "address": {
                            "text": "3 Second Hospital Avenue, #02-00 Health Promotion Board Building, Singapore 168937",
                            "country": "Singapore"
                        }
                    }
                },
                {
                    "resource": {
                        "resourceType": "Organization",
                        "id": "administeringCentre-1",
                        "name": "Outram Polyclinic",
                        "address": {
                            "text": "3 Second Hospital Avenue, #02-00 Health Promotion Board Building, Singapore 168937",
                            "country": "Singapore"
                        }
                    }
                }
            ]
        }
    },
    "holderDid": "did:elem:notExist"
}
```
</details>

Reply:
> It leads to fails during dependent lib compilation

But your link points to _test_ errors, not compilation errors?

>  Also, I can't manage find a way how to cover VCs with nested arrays

I just wrote some types according to your code implementation. It seems that all these service methods do not cover nested arrays either?

What are the acceptable values of `fields` in this case?

Reply:
> But your link points to _test_ errors, not compilation errors?  

seems wrong pipeline, this one https://github.com/affinityproject/affinidi-core-sdk/runs/4473328218?check_suite_focus=true

> I just wrote some types according to your code implementation. It seems that all these service methods do not cover nested arrays either?
Exactly, it should be covered by jsonld-signatures or, for now, we can write a disclaimer that we don't support picking nested fields in values of arrays. 

> What are the acceptable values of `fields` in this case?
Just anything that a developer decided to pass into this method. And the result could contain null fields. We are putting responsibility for selecting wrong fields on the developer who is using API. If you think it is better to find workaround how to add types that cover everything - let's do it in another PR as it is another story. Main points of this PR are:

Reply:
> But your link points to _test_ errors, not compilation errors?  

seems wrong pipeline, this one https://github.com/affinityproject/affinidi-core-sdk/runs/4473328218?check_suite_focus=true

> I just wrote some types according to your code implementation. It seems that all these service methods do not cover nested arrays either?
Exactly, it should be covered by jsonld-signatures or, for now, we can write a disclaimer that we don't support picking nested fields in values of arrays. 

> What are the acceptable values of `fields` in this case?
Just anything that a developer decided to pass into this method. And the result could contain null fields. We are putting responsibility for selecting wrong fields on the developer who is using API. If you think it is better to find workaround how to add types that cover everything - let's do it in another PR as it is another story. Main points of this PR are:
* [cover the issue with BBS signatures ](https://github.com/mattrglobal/node-bbs-signatures/issues/174)
* allow our users to select nested fields of their VCs

Reply:
> We are putting responsibility for selecting wrong fields on the developer who is using API.

But what are the _right_ fields in this case?

Reply:
it depends on what the VC owner decided to derive. You can find examples in this PR. 
I'll try to improve this, probably add some not strict typing and add tests for VCs with arrays. Let's make it step by step.

Reply:
> You can find examples in this PR.

But these examples do not contain any arrays?

> I'll try to improve this, probably add some not strict typing and add tests for VCs with arrays. Let's make it step by step.

So we are once again merging a PR with changes we do not fully understand ourselves?

If we don't know what are the correct possible values for VCs with arrays, how is VC owner supposed to know that?

Reply:
Yes, you are right.
But let's move step by step - these changes are really important as it introduces fixes for security vulnerability.
Also nested selective disclosure is stable for vcs that doesn't have arrays.
For this special case, we have a ticket https://lemmatree.atlassian.net/browse/NUC-180